dockerspawner/dockerspawner.py

```diff
@@ -1118,22 +1130,73 @@ def stop(self, now=False):
 
         self.clear_state()
 
-    def _volumes_to_binds(self, volumes, binds, mode="rw"):
+    @gen.coroutine
+    def ensure_volume(self, volume):
+        """Create a Docker volume with the configured driver, if it doesn't exist.
+        """
+        existing = yield self.docker('volumes')
+
+        if volume['source'] not in [v['Name'] for v in existing['Volumes']]:
+            self.log.info("Creating volume %s", volume['source'])
+            yield self.docker(
+                'create_volume',
+                name=volume['source'],
+                driver=volume.get('driver'),
+                driver_opts=volume.get('driver_opts')
+            )
+
+    def _volumes_to_binds(self, volumes, mode="rw"):
         """Extract the volume mount points from volumes property.
 
         Returns a dict of dict entries of the form::
 
             {'/host/dir': {'bind': '/guest/dir': 'mode': 'rw'}}
         """
+        return {
+            v['source']: {
+                'bind': v['target'], 
+                'mode': v.get('mode', mode)
+            }
+            for v in self._format_volumes(volumes)
+        }
+
+    def _format_volumes(self, volumes):
+        """Constructs a normalized representation of volumes from the volumes property.
+
+        The 'source' and 'target' are formatted using format_volume_name.
+
+        Returns a list of volumes of the form::
 
+            {
+                'source': volume_or_hostdir_name, 
+                'target': container_mountpoint,
+                'mode': readwrite_mode,
+                'driver': volume_driver,
+                'driver_opts': {**volume_driver_options},
+            }
+        """
         def _fmt(v):
             return self.format_volume_name(v, self)
 
+        formatted = []
         for k, v in volumes.items():
-            m = mode
             if isinstance(v, dict):
-                if "mode" in v:
-                    m = v["mode"]
-                v = v["bind"]
-            binds[_fmt(k)] = {"bind": _fmt(v), "mode": m}
-        return binds
+                v = copy(v)
+            else:
+                # support simple format like {source: target}
+                v = {'target': v}
+
+            if not 'target' in v:
```

Comment:
Err, looks like I forgot to keep supporting the format `{"bind": .., ..}`--will update.